crypto/perlasm/x86_64-xlate.pl

```diff
@@ -909,29 +909,54 @@
 		    # to deal with nasm/masm assembly.
 		    #
 		    $self->{value} =~ s/(.+)\s+align\s*=.*$/$1/;
+                    $current_segment = pop(@segment_stack);
+                    if (not $current_segment) {
+                        # if no previous section is defined, then assume .text
+                        # so code does not land in .data section by accident.
+                        # this deals with inconsistency of perl-assembly files.
+                        push(@segment_stack, ".text");
+                    }
 		    #
 		    # $$line may still contains align= option. We do care
 		    # about section type here.
 		    #
 		    $current_segment = $$line;
 		    $current_segment =~ s/([^\s]+).*$/$1/;
+                    push(@segment_stack, $current_segment);
 		    if (!$elf && $current_segment eq ".rodata") {
 			if	($flavour eq "macosx") { $self->{value} = ".section\t__DATA,__const"; }
+			elsif	($flavour eq "mingw64")	{ $self->{value} = ".section\t.rodata"; }
 		    }
 		    if (!$elf && $current_segment eq ".init") {
 			if	($flavour eq "macosx")	{ $self->{value} = ".mod_init_func"; }
 			elsif	($flavour eq "mingw64")	{ $self->{value} = ".section\t.ctors"; }
 		    }
 		} elsif ($dir =~ /\.(text|data)/) {
+                    $current_segment = pop(@segment_stack);
 		    $current_segment=".$1";
+                    if (not $current_segment) {
+                        # if no previous section is defined, then assume .text
+                        # so code does not land in .data section by accident.
+                        # this deals with inconsistency of perl-assembly files.
+                        push(@segment_stack, ".text");
+                    }
```

Comment:
This does not make much sense to me here - it cannot happen as there will be either text or data from the regexp match. Maybe the ordering is wrong?